lib/cloud-operation/communication/communication-channel-base.ts

```diff
@@ -0,0 +1,120 @@
+import { EventEmitter } from "events";
+import { CloudOperationMessageTypes, CloudCommunicationChannelExitCodes, CloudCommunicationEvents } from "../../constants";
+
+export abstract class CommunicationChannelBase extends EventEmitter implements ICloudCommunicationChannel {
+	private static readonly PING_INTERVAL: number = 10000;
+	private static readonly MISSING_PING_REPLIES_COUNT: number = 6;
+
+	protected isConnected: boolean;
+	private pingInterval: NodeJS.Timer;
+	private echoSequenceNumber: number = 0;
+	private echoReplySequenceNumber: number = 0;
+
+	private handshakeCompleteResolve: (value?: PromiseLike<void>) => void;
+
+	constructor(protected cloudOperationId: string,
+		protected data: ICloudCommunicationChannelData<any>,
+		protected $logger: ILogger) {
+		super();
+	}
+
+	public async connect(): Promise<void> {
+		if (!this.isConnected) {
+			await this.connectCore();
```

Comment:
connectCore can be rejected - do we need try .. catch?

Reply:
No. We want to throw the error.